### PR TITLE
chore: migrate to pnpm

### PR DESCRIPTION
**Pull request by Jinjiang**
... and

- fixed some typescript errors
- improved some test cases

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -10,35 +10,44 @@ jobs:
...
-      - name: Set node version to 16
-        uses: actions/setup-node@v2
+      - uses: actions/checkout@v3
+      - uses: pnpm/action-setup@v2
         with:
-          node-version: 16
-          cache: 'yarn'
-      - run: yarn install
-      - run: yarn test:webpack4
+          version: 8.12.0
```
Comment (sxzz) [inline .github/workflows/ci.yml:16]: Let's add `packageManager` field into `package.json` and remove this.

Comment (sxzz) [reply .github/workflows/ci.yml:16]: Same for below

Comment (Jinjiang) [reply .github/workflows/ci.yml]: done

**Review by sodatea — APPROVED** [review]

**Comment by Airkro** [comment]
Using pnpm 8.x, you should add this file:

```ini
# .npmrc
enable-pre-post-scripts = true
```

to make sure run `pretest` before `test`.

Or using pnpm 9.x

See: 

- https://pnpm.io/8.x/cli/run#npmrc-settings
- https://pnpm.io/cli/run#npmrc-settings

**Pull request merged**